### ISSUE TITLE
Add fully convolutional cnns for scene segmentation

### ISSUE THREAD

**Issue by agibsonccc** [deeplearning4j/deeplearning4j#2818]
https://people.eecs.berkeley.edu/~jonlong/long_shelhamer_fcn.pdf

**Comment by evariste** [comment]
A description of the problem/use case: Assuming an image as input with a corresponding per-pixel labels (e.g. `cat/not-cat`). Imagine a (very poor) network that simply applies a convolution layer (CL) to the image and has two output channels + softmax activation. User seeks to evaluate the loss directly from CL output over all the per-pixel labels.

Existing output layer in dl4j by default applies a full set of connections between their left-hand input and the loss function. In the case of semantic segmentation, this is not necessary. What is needed instead is either an output layer that takes the activations of a convolutional layer and directly applies the loss function to each pixel,  or an convolutional layer that can also act as an output layer.

Hope this makes sense.

**Comment by turambar** [comment]
Just wanted to comment on this: the more general case of FCNs requires the ability to handle inputs of any size. Currently DL4J requires input size to be fixed ahead of time.

**Comment by evariste** [comment]
I attach some classes as an initial solution and code for a test executable. I have also attached train/test data separately.

This code uses versions 0.7.2 of DL4j and ND4j and requires a fix to ND4j that we proposed separately to be included ([link](https://github.com/deeplearning4j/nd4j/issues/1611)) othewise the labels are not used correctly during training.

[issue_2818_src.zip](https://github.com/deeplearning4j/deeplearning4j/files/796351/issue_2818_src.zip)

[issue_2818_test_data.zip](https://github.com/deeplearning4j/deeplearning4j/files/796354/issue_2818_test_data.zip)

**Comment by evariste** [comment]
@turambar , yes, there is an issue regarding sizing. AFAIK, if input dimensions can be determined from the data, then, e.g., `GraphBuilder.SetInputTypes` can be used on the way in. The output dimensions are more tricky as they depend on the number of convolutions applied, kernel size, stride, mode etc. In the source described above it is just hard-coded.

**Comment by AlexDBlack** [comment]
I believe we have everything we need now for segmentation, with the addition of CnnLossLayer

**Comment by AlexDBlack** [comment]
@zibellon no examples (but there are unit tests: https://github.com/deeplearning4j/deeplearning4j/search?q=CnnLossLayer&unscoped_q=CnnLossLayer )

There's not much to it. It goes after a CNN layer, and really just needs a loss function and activation function.
Oh, and for future reference: ask in gitter for these sorts of questions, not closed issues.
